File: leetcode_300_Longest Increasing Subsequence_O(nlogn).cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp;
        for(int i=0;i<nums.size();i++){
            int right = dp.size();
            int left = 0;
            while(left<right){
                int mid = left + (right-left)/2;
                if(dp[mid]<nums[i]) left = mid + 1;
                else right = mid;
            }
            if(right>=dp.size()) dp.push_back(nums[i]);
            else dp[right] = nums[i];
        }
        return dp.size();
    }
};

File: leetcode_300_Longest Increasing Subsequence_O(nlogn).py
def lengthOfLIS(self, nums):
    dp = [0] * len(nums)
    size = 0
    for x in nums:
        i, j = 0, size
        while i != j:
            m = (i + j) / 2
            if dp[m] < x:
                i = m + 1
            else:
                j = m
        dp[i] = x
        size = max(i + 1, size)
    return size